Source/JavaScriptCore/b3/B3ReduceStrength.cpp

```diff
@@ -3405,6 +3411,110 @@ class ReduceStrength {
         return false;
     }
 
+    // For Op==BitOr or BitXor, turn any of these:
+    //      Op(Shl(x1, x2), ZShr(x1, 32|64-x2))
+    //      Op(Shl(x1, x2), ZShr(x1, Sub(32|64, x2)))
+    //      Op(ZShr(x1, 32|64-x2), Shl(x1, x2))
+    //      Op(ZShr(x1, Sub(32|64, x2)), Shl(x1, x2))
+    // Into this: RotL(x1, x2)
+    // And any of these:
+    //      Op(ZShr(x1, x2), Shl(x1, 32|64-x2))
+    //      Op(ZShr(x1, x2), Shl(x1, Sub(32|64, x2)))
+    //      Op(Shl(x1, 32|64-x2), ZShr(x1, x2))
+    //      Op(Shl(x1, Sub(32|64, x2)), ZShr(x1, x2))
+    // Into this: RotR(x1, x2)
+    // This pattern is widely used to represent bitwise rotation in languages that don't support it natively,
+    // and appears in hot loops in cryptography algorithms, so we should try to match it when possible.
+    bool handleBitRotation()
+    {
+        if ((m_value->child(0)->opcode() == Shl
+                && m_value->child(1)->opcode() == ZShr)
+            || (m_value->child(0)->opcode() == ZShr
+                && m_value->child(1)->opcode() == Shl)) {
+            Value* shl = m_value->child(0)->opcode() == Shl ? m_value->child(0) : m_value->child(1);
+            Value* shr = m_value->child(0)->opcode() == ZShr ? m_value->child(0) : m_value->child(1);
+
+            // Handle the cases where both shifts have a constant shift amount.
+            if (m_value->type() == Int32
+                && shl->child(1)->hasInt()
+                && shr->child(1)->hasInt()
+                && shl->child(1)->asInt() + shr->child(1)->asInt() == 32
```

Comment:
We should check both are positive since `-1 + 33 == 32`.

Reply:
B3 assumes the shift amount is modulo the number of bits, so in the case we have `-1` and `33`, then `-1 mod 32 = 31` and `33 mod 32 = 1` and `31` and `1` are a valid pair to apply this optimization to. So I think this is fine.

We should definitely check if the sum overflows though, good catch.